habitat_baselines/rl/ppo/ppo_trainer.py

```diff
@@ -0,0 +1,519 @@
+            n_envs = self.envs.num_envs
+            for i in range(n_envs):
+                if (
+                    next_episodes[i].scene_id,
+                    next_episodes[i].episode_id,
+                ) in stats_episodes:
+                    envs_to_pause.append(i)
+
+                # episode ended
+                if not_done_masks[i].item() == 0:
+                    episode_stats = dict()
+                    episode_stats["spl"] = infos[i]["spl"]
+                    episode_stats["success"] = int(infos[i]["spl"] > 0)
+                    episode_stats["reward"] = current_episode_reward[i].item()
+                    current_episode_reward[i] = 0
+                    # use scene_id + episode_id as unique id for storing stats
+                    stats_episodes[
+                        (
+                            current_episodes[i].scene_id,
+                            current_episodes[i].episode_id,
+                        )
+                    ] = episode_stats
+                    if args.video_option:
+                        generate_video(
+                            args,
+                            rgb_frames[i],
+                            current_episodes[i].episode_id,
+                            cur_ckpt_idx,
+                            infos[i]["spl"],
+                            writer,
+                        )
+                        rgb_frames[i] = []
+
+                # episode continues
+                elif args.video_option:
+                    frame = observations_to_image(observations[i], infos[i])
+                    rgb_frames[i].append(frame)
+
+            # pausing self.envs with no new episode
+            if len(envs_to_pause) > 0:
+                state_index = list(range(self.envs.num_envs))
+                for idx in reversed(envs_to_pause):
+                    state_index.pop(idx)
+                    self.envs.pause_at(idx)
+
+                # indexing along the batch dimensions
+                test_recurrent_hidden_states = test_recurrent_hidden_states[
+                    state_index
+                ]
+                not_done_masks = not_done_masks[state_index]
+                current_episode_reward = current_episode_reward[state_index]
+
+                for k, v in batch.items():
+                    batch[k] = v[state_index]
+
+                if args.video_option:
+                    rgb_frames = [rgb_frames[i] for i in state_index]
+
+        aggregated_stats = dict()
+        for stat_key in next(iter(stats_episodes.values())).keys():
+            aggregated_stats[stat_key] = sum(
+                [v[stat_key] for v in stats_episodes.values()]
+            )
+        num_episodes = len(stats_episodes)
+
+        episode_reward_mean = aggregated_stats["reward"] / num_episodes
+        episode_spl_mean = aggregated_stats["spl"] / num_episodes
+        episode_success_mean = aggregated_stats["success"] / num_episodes
+
+        logger.info(
```

Comment:
The logging and writer accesses below should be abstracted out to a separate util method.